handlers/PublicHandlers.py

```diff
@@ -61,16 +65,149 @@ def get(self, *args, **kwargs):
+        hasAdminRole = self.is_admin(claims)
+
+        # Get the team code (if set) would have come from the Join Team page.
+        team_code = None;
+        if "teamcode" in code_flow:
+            team_code = code_flow["teamcode"]
+
+        # Look up user the Azure AD object id (Guid)
+        user = User.by_uuid(claims["oid"])
+
+        # If user is not admin and not part of a team or new, they need to join a team.
+        # Redirect to the join team page here, if needed, before the user is created in the db.
+        if self.needs_team(hasAdminRole, user, team_code):
+            self.redirect("/jointeam?upn=" + claims["upn"])
+            return
+
+        # New user
+        if user is None:
+            user = self.add_user(claims, team_code)
+
+        self.update_permissions(user, hasAdminRole)
+
+        self.dbsession.commit()
+
+        self.create_login_session(user)
+
+        self.redirect("/user")
+
+    def create_login_session(self, user):
+        self.start_session()
+        theme = Theme.by_id(user.theme_id)
+        if user.team is not None:
+            self.session["team_id"] = int(user.team.id)
+        self.session["user_id"] = int(user.id)
+        self.session["user_uuid"] = user.uuid
+        self.session["handle"] = user.handle
+        self.session["theme"] = [str(f) for f in theme.files]
+        self.session["theme_id"] = int(theme.id)
+        if user.is_admin():
+            self.session["menu"] = "admin"
+        else:
+            self.session["menu"] = "user"
+        self.session.save()
+
+    # Admin is defined by the Azure AD AppRole "Admin"
+    def is_admin(self, claims):
+        if "roles" in claims:
+            roles = claims["roles"]
+            if "Admin" in roles:
+                return True
+        return False
+
+    # Figure out whether the user should be presented with the join team page.
+    def needs_team(self, hasAdminRole, user, team_code):
+        needsateam = False
+        # Is the a member of a team, if not admin?
+        if not hasAdminRole:
+            if user is not None:
+                if user.team is None:
+                    # Have they entered a joining code from the join team page?
+                    if team_code is not None and len(team_code) > 0:
+                        team = Team.by_code(team_code)
+                        user.team = team
+                    else:
+                        needsateam = True
+            # New user must have teamcode for joining a team.
+            elif team_code is None or len(team_code) == 0:
+                needsateam = True
+        return needsateam
+
+    def add_user(self, claims, team_code):
+        user = User()
+        user.uuid = claims["oid"]
+        user.handle = claims["preferred_username"].split("@")[0]
+        # Generate a long random password that the user will never know or use.
+        user.password = ''.join(random.choices(string.ascii_letters + string.digits + string.punctuation, k=30))
```

Comment:
Very minor nitpick -- I would encourage the use of [`secrets`](https://docs.python.org/3/library/secrets.html?highlight=secrets#module-secrets) or [`SystemRandom`](https://docs.python.org/3/library/random.html#random.SystemRandom) here for password generation as `random` is not cryptographically secure.

Reply:
Changed to use `secrets`.